pkg/middleware/trace.go

```diff
@@ -0,0 +1,36 @@
+/****************************************************************************
+ * Copyright 2023 Optimizely, Inc. and contributors                        *
+ *                                                                          *
+ * Licensed under the Apache License, Version 2.0 (the "License");          *
+ * you may not use this file except in compliance with the License.         *
+ * You may obtain a copy of the License at                                  *
+ *                                                                          *
+ *    http://www.apache.org/licenses/LICENSE-2.0                            *
+ *                                                                          *
+ * Unless required by applicable law or agreed to in writing, software      *
+ * distributed under the License is distributed on an "AS IS" BASIS,        *
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. *
+ * See the License for the specific language governing permissions and      *
+ * limitations under the License.                                           *
+ ***************************************************************************/
+
+// Package middleware //
+package middleware
+
+import (
+	"net/http"
+
+	"go.opentelemetry.io/otel"
+)
+
+func AddTracing(tracerName, spanName string) func(http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		fn := func(w http.ResponseWriter, r *http.Request) {
+			ctx, span := otel.Tracer(tracerName).Start(r.Context(), spanName)
+			defer span.End()
+
+			next.ServeHTTP(w, r.WithContext(ctx))
```

Comment:
I do not see how to extract or add trace id for distributed tracing. Can we add it here?

Reply:
It seems, every trace (in this case every API call) has (should be) a different trace id and it is the ideal otherwise each trace can't be identified separately.

Reply:
My question is that if upstream services add trace-id to agent requests, we need to extract it and add to our traces.

Reply:
The OpenTelemetry itself generates a random number and sets it as trace id actually.